pkg/providers/amifamily/types.go

```diff
@@ -102,9 +123,10 @@ type DescribeImageQuery struct {
 func (q DescribeImageQuery) DescribeImagesInput() *ec2.DescribeImagesInput {
 	return &ec2.DescribeImagesInput{
 		// Don't include filters in the Describe Images call as EC2 API doesn't allow empty filters.
-		Filters:    lo.Ternary(len(q.Filters) > 0, q.Filters, nil),
-		Owners:     lo.Ternary(len(q.Owners) > 0, lo.ToSlicePtr(q.Owners), nil),
-		MaxResults: aws.Int64(1000),
+		Filters:           lo.Ternary(len(q.Filters) > 0, q.Filters, nil),
+		Owners:            lo.Ternary(len(q.Owners) > 0, lo.ToSlicePtr(q.Owners), nil),
+		IncludeDeprecated: aws.Bool(true),
+		MaxResults:        aws.Int64(1000),
```

Comment:
Also, what is the expectation around drift? Currenlty, karpenter will use the resolved list of AMIs to check against the instances that were launched by Karpenter. If the AMI used on a node exist in the resolved list of AMIs the node will not be drifted.  With this PR, if an AMI is deprecated and a new AMI is resolved by the same `amiSelectorTerms`, the existing nodes would not be drifted however new nodes launched will use the new non-deprecated AMI. What is thinking here? Should we drift nodes that are using deprecated AMIs if there is a new non-deprecated AMI is resolved?

Reply:
In general the idea was when following the pattern as described here https://karpenter.sh/docs/tasks/managing-amis/#option-2-lock-down-which-amis-are-selected that Karpenter should not disrupt workloads when an AMI is deprecated and `amiSelectorTerms` contain only an `id`. Karpenter does a real amazing job of detecting newer amis when launching nodes when `amiSelectorTerms` contains names or tags.

> Should we drift nodes that are using deprecated AMIs if there is a new non-deprecated AMI is resolved?

In my opinion yes if a new non-deprecated AMI is available nodes should be drifted. I think that would warrant a change [here](https://github.com/aws/karpenter-provider-aws/blob/main/pkg/providers/amifamily/ami.go#L171)

Reply:
IIRC our previous conversations decided:
1. We will migrate off of deprecated AMIs if we have multiple in the selector
2. We will keep using deprecated AMIs if there is no non-deprecated alternative

Reply:
Okay, Karpenter will scale-up nodes based on the sorted AMIs, but not for drift. We only check if the instance AMI is within the resolved list of AMIs for the nodeclass. We will need to update our drift detection to find nodes that are using deprecated AMIs that can be update to drifted to us non-deprecated. 

> Karpenter does a real amazing job of detecting newer amis when launching nodes when amiSelectorTerms contains names or tags which support wildcards.

> 1. We will migrate off of deprecated AMIs if we have multiple in the selector

In the current state of the PR, we will need to make updates [here](https://github.com/aws/karpenter-provider-aws/blob/9d8849fa0a2634df11bdccd5a22782c2640c3355/pkg/cloudprovider/drift.go#L72) or we break that behavior all together.

Reply:
I think this decision might require broader review from the other maintainers.

Reply:
@ellistarn @engedaam my apologies for the delayed response but I want to share my findings and make sure I am not misunderstanding the requirement and discussion here.

1. `IncludeDeprecated` only takes effect when the AMI owner is not self (the current account) for example when a central AWS account is responsible for building and sharing AMIs to multiple accounts in the organization.  
2. I ran a few tests on my sandbox EKS cluster and the drift works as expected as long as a new non deprecated AMI is available and the `amiSelectorTerms` don't have the AMI id pinned. As an example here's what i tested with (as an example with nodepool of type of only 1 `instance-family` and `arch` 
```
spec:
  amiFamily: AL2
  amiSelectorTerms:
  - name: dummy-amazon-eks-node-1.29-*
    owner: "111111111111" 
```
The above is in AWS account `999999999999` and the AMI is being shared from AWS account `111111111111`

3. Next due to the selector terms we see that all nodes from this ec2nc are using the AMI 
```
status:
  amis:
  - id: ami-053ccf7a5d1c364ac
    name: dummy-amazon-eks-node-1.29-test-20240625
    requirements:
    - key: kubernetes.io/arch
      operator: In
      values:
      - amd64
  conditions:
  - lastTransitionTime: "2024-07-22T22:43:42Z"
    message: ""
    reason: Ready
    status: "True"
    type: Ready
 ```
 
 and the node claim 
 ```
   conditions:
  - lastTransitionTime: "2024-07-25T04:18:14Z"
    message: ""
    reason: Initialized
    status: "True"
    type: Initialized
  - lastTransitionTime: "2024-07-25T04:17:12Z"
    message: ""
    reason: Launched
    status: "True"
    type: Launched
  - lastTransitionTime: "2024-07-25T04:18:14Z"
    message: ""
    reason: Ready
    status: "True"
    type: Ready
  - lastTransitionTime: "2024-07-25T04:18:06Z"
    message: ""
    reason: Registered
    status: "True"
    type: Registered
  imageID: ami-053ccf7a5d1c364ac
  nodeName: ip-172-xx-xxx-xx.us-west-2.compute.internal
  providerID: aws:///us-west-2b/i-0xxxxxxxxxxxxxx82
  ```

Reply:
@ellistarn @engedaam my apologies for the delayed response but I want to share my findings and make sure I am not misunderstanding the requirement and discussion here.

1. `IncludeDeprecated` only takes effect when the AMI owner is not self (the current account) for example when a central AWS account is responsible for building and sharing AMIs to multiple accounts in the organization.  By default deprecated AMIs in the same AWS account i.e the owner is self are visible always.
2. I ran a few tests on my sandbox EKS cluster and the drift works as expected as long as a new non deprecated AMI is available and the `amiSelectorTerms` don't have the AMI id pinned. As an example here's what i tested with (as an example with nodepool of type of only 1 `instance-family` and `arch` ) 

I ran the same tests below with using tags and name on the `amiSelectorTerms`

```
spec:
  amiFamily: AL2
  amiSelectorTerms:
  - name: dummy-amazon-eks-node-1.29-*
    owner: "111111111111" 
```
The above is in AWS account `999999999999` and the AMI is being shared from AWS account `111111111111`

3. Next due to the selector terms we see that all nodes from this ec2nc are using the AMI 
```
status:
  amis:
  - id: ami-053ccf7a5d1c364ac
    name: dummy-amazon-eks-node-1.29-test-20240625
    requirements:
    - key: kubernetes.io/arch
      operator: In
      values:
      - amd64
```
 
and the node claim 
```
   conditions:
  - lastTransitionTime: "2024-07-25T04:18:14Z"
    message: ""
    reason: Initialized
    status: "True"
    type: Initialized
  - lastTransitionTime: "2024-07-25T04:17:12Z"
    message: ""
    reason: Launched
    status: "True"
    type: Launched
  - lastTransitionTime: "2024-07-25T04:18:14Z"
    message: ""
    reason: Ready
    status: "True"
    type: Ready
  - lastTransitionTime: "2024-07-25T04:18:06Z"
    message: ""
    reason: Registered
    status: "True"
    type: Registered
  imageID: ami-053ccf7a5d1c364ac
  nodeName: ip-172-xx-xxx-xx.us-west-2.compute.internal
  providerID: aws:///us-west-2b/i-0xxxxxxxxxxxxxx82
```
4. I then deprecated this AMI as visible below and scaled up pods in an namespace which scaled up and added new nodeclaims and worked as expected - otherwise without the changes in this PR Karpenter would fail to scale up new nodes

```
✗ aws ec2 describe-images --region us-west-2 --image-ids ami-053ccf7a5d1c364ac
{
    "Images": [
        {
            "Architecture": "x86_64",
            "CreationDate": "2024-07-25T04:07:58.000Z",
            "ImageId": "ami-053ccf7a5d1c364ac",
            "ImageLocation": "111111111111/dummy-amazon-eks-node-1.29-test-20240625",
            "ImageType": "machine",
            "Public": false,
            "OwnerId": "111111111111",
            "PlatformDetails": "Linux/UNIX",
            "UsageOperation": "RunInstances",
            "State": "available",
            ...
            ...
            ...
            ...
            "VirtualizationType": "hvm",
            "DeprecationTime": "2024-07-25T22:30:00.000Z",
            "ImdsSupport": "v2.0"
        }
    ]
}
```
5. Then, I created a new AMI in AWS account `111111111111` and shared it with account `999999999999`. Since the creationDate of this new AMI was more recent - I see the drift working as expected.

Here's the nodeclaims from the new AMI
```
  conditions:
  - lastTransitionTime: "2024-07-29T21:49:52Z"
    message: ""
    reason: Initialized
    status: "True"
    type: Initialized
  - lastTransitionTime: "2024-07-29T21:48:58Z"
    message: ""
    reason: Launched
    status: "True"
    type: Launched
  - lastTransitionTime: "2024-07-29T21:49:52Z"
    message: ""
    reason: Ready
    status: "True"
    type: Ready
  - lastTransitionTime: "2024-07-29T21:49:44Z"
    message: ""
    reason: Registered
    status: "True"
    type: Registered
  imageID: ami-0ab3fd8c11a5baed9
  nodeName: ip-172-xx-xxx-xxx.us-west-2.compute.internal
  providerID: aws:///us-west-2c/i-0xxxxxxxxxxxxxxdc
```

and the ec2nc
```
status:
  amis:
  - id: ami-0ab3fd8c11a5baed9
    name: dummy-amazon-eks-node-1.29-test-20240703
    requirements:
    - key: kubernetes.io/arch
      operator: In
      values:
      - amd64
```

and the nodeclaims 
```
→ [04:50:17] (git:main *=) ✗ k get nodeclaim
NAME                  TYPE         ZONE         NODE                                           READY   AGE
cosmic-lamb-1-5kxj7   m5.xlarge    us-west-2c   ip-172-xx-xxx-xxx.us-west-2.compute.internal   True    3m27s
cosmic-lamb-1-cqjcm   m5.xlarge    us-west-2a   ip-172-xx-xxx-xxx.us-west-2.compute.internal   True    4d17h
cosmic-lamb-1-fccr4   m5.xlarge    us-west-2b   ip-172-xx-xxx-xxx.us-west-2.compute.internal   True    4m39s
→ [04:52:23] (git:main *=) ✗ k get nodeclaim
NAME                  TYPE         ZONE         NODE                                           READY   AGE
cosmic-lamb-1-5kxj7   m5.xlarge    us-west-2c   ip-172-xx-xxx-xxx.us-west-2.compute.internal   True    3m48s
cosmic-lamb-1-fccr4   m5.xlarge    us-west-2b   ip-172-xx-xxx-xxx.us-west-2.compute.internal   True    5m
```

and the new AMI 

```
✗ aws ec2 describe-images --region us-west-2 --image-ids ami-0ab3fd8c11a5baed9 --profile vdevicepop-dev-admin
{
    "Images": [
        {
            "Architecture": "x86_64",
            "CreationDate": "2024-07-29T21:16:39.000Z",
            "ImageId": "ami-0ab3fd8c11a5baed9",
            "ImageLocation": "111111111111/dummy-amazon-eks-node-1.29-test-20240703",
            "ImageType": "machine",
            "Public": false,
            "OwnerId": "111111111111",
            "PlatformDetails": "Linux/UNIX",
            "UsageOperation": "RunInstances",
            "State": "available",
            ...
            ...
            ...
            ...
            "EnaSupport": true,
            "Hypervisor": "xen",
            "Name": "dummy-amazon-eks-node-1.29-test-20240703",
            "RootDeviceName": "/dev/xvda",
            "RootDeviceType": "ebs",
            "SriovNetSupport": "simple",
            "VirtualizationType": "hvm",
            "ImdsSupport": "v2.0"
        }
    ]
}
```

Reply:
https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/disable-an-ami.html 

When an AMI is disabled - there's a separate flag for that `IncludeDisabled` 

Additionally

> A disabled AMI can't be shared. If an AMI was public or previously shared, it is made private. If an AMI was shared with an AWS account, organization, or Organizational Unit, they lose access to the disabled AMI.

and 

>EC2 instances that were previously launched using an AMI that is subsequently disabled are not affected, and can be stopped, started, and rebooted.

Reply:
@ellistarn @engedaam @jonathan-innis awesome 🥳  karpenter v1 is released, i am excited to try it out soon. I just wanted to check with you all if we can reinitiate the conversation for this PR?

Reply:
@engedaam @jonathan-innis apologies for the tag but just wanted to bump this PR again and see if we can move forward.

Reply:
> The only scenario where drift doesn't work is when an AMI is deprecated and no new AMIs are available, Karpenter will not fall back to a non deprecated AMI with an older creationDate from the deprecated AMI.

I think that is what we want. If a there are any alternatives, Karpenter should use those AMIs. That makes sense to me. I think we can close out this comment then. @shabbskagalwala can you an integration test with that we expect with drift and AMI selection?

Reply:
Done, have updated the PR with the latest changes. 

* **Pending tests** i'll have those out soon